web/client-api/src/main/java/io/deephaven/ide/shared/IdeSession.java

```diff
@@ -165,39 +166,137 @@ public JsRunnable onLogMessage(JsConsumer<LogItem> callback) {
-        throw new UnsupportedOperationException("changeDocument");
+        // translate arbitrary value from js to our "properly typed request object".
+        final JsPropertyMap<Object> jsMap = (JsPropertyMap<Object>) params;
+        final ChangeDocumentRequest request = new ChangeDocumentRequest();
+        request.setConsoleId(result);
+        final JsPropertyMap<Object> textDoc = jsMap.getAny("textDocument").asPropertyMap();
+        final VersionedTextDocumentIdentifier textDocument = new VersionedTextDocumentIdentifier();
+        textDocument.setUri(textDoc.getAny("uri").asString());
+        textDocument.setVersion(textDoc.getAny("version").asDouble());
+        request.setTextDocument(textDocument);
+
+        final JsArrayLike<Object> changes = jsMap.getAny("contentChanges").asArrayLike();
+        final JsArray<TextDocumentContentChangeEvent> changeList = new JsArray<>();
+        for (int i = 0; i < changes.getLength(); i ++ ) {
+            final JsPropertyMap<Object> change = changes.getAnyAt(i).asPropertyMap();
+            final TextDocumentContentChangeEvent changeItem = new TextDocumentContentChangeEvent();
+            changeItem.setText(change.getAny("text").asString());
+            if (change.has("rangeLength")) {
+                changeItem.setRangeLength(change.getAny("rangeLength").asInt());
+            }
+            if (change.has("range")) {
+                changeItem.setRange(toRange(change.getAny("range")));
+            }
+            changeList.push(changeItem);
+        }
+        request.setContentChangesList(changeList);
+
+        JsLog.debug("Sending content changes", request);
+        connection.consoleServiceClient().changeDocument(request, connection.metadata(),
+                (p0, p1) -> JsLog.debug("Updated doc", p0, p1));
     }
 
-    public Promise<JsArray<CompletionItem>> getCompletionItems(Object params) {
-        LazyPromise<CompletionItem[]> promise = new LazyPromise<>();
+    private DocumentRange toRange(final Any range) {
+        final JsPropertyMap<Object> rangeObj = range.asPropertyMap();
+        final DocumentRange result = new DocumentRange();
+        result.setStart(toPosition(rangeObj.getAny("start")));
+        result.setEnd(toPosition(rangeObj.getAny("end")));
+        return result;
+    }
 
-        throw new UnsupportedOperationException("getCompletionItems");
-//        connection.getServer().getCompletionItems(result.getHandle(), new CompletionParams((JsPropertyMap<Object>)params), promise.asCallback());
-//
-//        return promise.asPromise(BatchOperation.MAX_BATCH_TIME)
-//                .then(items->Promise.resolve(makeJsArrayFromCompletionItems(items)));
+    private Position toPosition(final Any pos) {
+        final JsPropertyMap<Object> posObj = pos.asPropertyMap();
+        final Position result = new Position();
+        result.setLine(posObj.getAny("line").asInt());
+        result.setCharacter(posObj.getAny("character").asInt());
```

Comment:
i may revisit these later and make jsinterop "beans" for them, since asInt() has no power under prod builds, and will just write garbage - we will want/need something that even at runtime in prod gets typechecked for calls from non-typesafe code

Reply:
luckily for us, these objects that come in from react will always be populated, so fine enough to leave like this for now.